sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/grouping.scala

```diff
@@ -155,3 +155,21 @@ object GroupingID {
     if (SQLConf.get.integerGroupingIdEnabled) IntegerType else LongType
   }
 }
+
+/**
+ * Wrapper expression to avoid further optimizations between the parent and child of this
+ * expression.
+ */
+case class GroupingExpression(child: Expression) extends UnaryExpression {
```

Comment:
This is a tagging-like expr (`TaggingExpression`) ?

Reply:
Thanks, I think it is. Changed in https://github.com/apache/spark/pull/31913/commits/6e05f14c1a21e2894c1a4a85546a790189a43f5a